Reducing the number of call to THIS

rkothiya opened the pull request:
The macro THIS calls a function.
Modified all the functions where THIS is called
multiple times by saving the return value of THIS
and reusing it.

Updates: #1683

Change-Id: I91328eb4af905ee4f8c9ff993dd0e734ade0183a
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
Reducing the number of call to THIS

The macro THIS calls a function.
Modified all the functions where THIS is called
multiple times by saving the return value of THIS
and reusing it.

Updates: #1683

Change-Id: I91328eb4af905ee4f8c9ff993dd0e734ade0183a
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
As mention in this issue there are 150 file where we need to change this. Instead of changing everything at once, I am trying to 
change few files at a time which will make the review process easier. 

I have not changed some functions ( for example inode_grep() ). As "THIS"  is called only once here, even after making the change, it will still be called once hence there was no point in doing that, so I skipped it.

mykaul commented on api/src/glfs-fops.c:
```diff
@@ -109,17 +109,18 @@ glfs_get_upcall_cache_invalidation(struct gf_upcall *to_up_data,
-    GF_VALIDATE_OR_GOTO(THIS->name, to_up_data, out);
-    GF_VALIDATE_OR_GOTO(THIS->name, from_up_data, out);
+    GF_VALIDATE_OR_GOTO(this->name, to_up_data, out);
+    GF_VALIDATE_OR_GOTO(this->name, from_up_data, out);
```
this check is not needed. This function needs to be static, and it's called from glfs_enqueue_upcall_data() only, which makes sure this variable is not null.
I'm in favor of removing ineffective validations first.

xhernandez replied:
I agree on removing the check. I also think that in this case we can change the function signature to include `this` to avoid using `THIS`.

rkothiya replied:
done

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -145,16 +146,18 @@ glfs_get_upcall_lease(struct gf_upcall *to_up_data,
 
-    GF_VALIDATE_OR_GOTO(THIS->name, to_up_data, out);
-    GF_VALIDATE_OR_GOTO(THIS->name, from_up_data, out);
+    xlator_t *this = THIS;
```
Here we can also pass `this` as an argument.

rkothiya replied:
done

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -1513,6 +1516,7 @@ glfs_pwritev_common(struct glfs_fd *glfd, const struct iovec *iovec, int iovcnt,
                     struct glfs_stat *poststat)
 {
     xlator_t *subvol = NULL;
+    xlator_t *this = THIS;
```
`THIS` at this point will likely be invalid. It cannot be used until it has been correctly initialized after line 1537.

This also happens in many other functions.

rkothiya replied:
Done

xhernandez reviewed (changes requested):
Almost all (if not all) the usages of `THIS` are just for logging purposes. I wonder if it's really important to use `this->name` instead of a constant value like `"gfapi"`.

Commit:
Reducing the number of call to THIS

Changes done to incorporate review comments.
1) Removed check GF_VALIDATE_OR_GOTO
2) Passed "this" to function where possible
3) Made some functions static

Updates: #1683

Change-Id: I9a22dc6c13c9859a955445fc742de45c23d124c9
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/api/src/glfs-fops.c b/api/src/glfs-fops.c
index 12374003a..5e700a6eb 100644
--- a/api/src/glfs-fops.c
+++ b/api/src/glfs-fops.c
@@ -104,7 +104,8 @@ glfd_set_state_bind(struct glfs_fd *glfd)
  */
 static int
 glfs_get_upcall_cache_invalidation(struct gf_upcall *to_up_data,
-                                   struct gf_upcall *from_up_data, xlator_t *this)
+                                   struct gf_upcall *from_up_data,
+                                   xlator_t *this)
 {
     struct gf_upcall_cache_invalidation *ca_data = NULL;
     struct gf_upcall_cache_invalidation *f_ca_data = NULL;
@@ -141,8 +142,6 @@ glfs_get_upcall_lease(struct gf_upcall *to_up_data,
     struct gf_upcall_recall_lease *f_ca_data = NULL;
     int ret = -1;
 
-
-
     f_ca_data = from_up_data->data;
 
     ca_data = GF_CALLOC(1, sizeof(*ca_data), glfs_mt_upcall_entry_t);
@@ -5392,7 +5391,6 @@ pub_glfs_fd_set_lkowner(struct glfs_fd *glfd, void *data, int len)
         goto invalid_fs;
     }
 
-
     if ((len <= 0) || (len > GFAPI_MAX_LOCK_OWNER_LEN)) {
         errno = EINVAL;
         gf_smsg(this->name, GF_LOG_ERROR, errno, API_MSG_INVALID_ARG,
@@ -5467,7 +5465,8 @@ invalid_fs:
 }
 
 static void
-glfs_enqueue_upcall_data(struct glfs *fs, struct gf_upcall *upcall_data, xlator_t *this)
+glfs_enqueue_upcall_data(struct glfs *fs, struct gf_upcall *upcall_data,
+                         xlator_t *this)
 {
     int ret = -1;
     upcall_entry *u_list = NULL;
@@ -5494,7 +5493,8 @@ glfs_enqueue_upcall_data(struct glfs *fs, struct gf_upcall *upcall_data, xlator_
                                                      upcall_data, this);
             break;
         case GF_UPCALL_RECALL_LEASE:
-            ret = glfs_get_upcall_lease(&u_list->upcall_data, upcall_data, this);
+            ret = glfs_get_upcall_lease(&u_list->upcall_data, upcall_data,
+                                        this);
             break;
         default:
             break;
@@ -6228,7 +6228,6 @@ pub_glfs_xreaddirplus_r(struct glfs_fd *glfd, uint32_t flags,
 
     GF_REF_GET(glfd);
 
-
     errno = 0;
 
     if (ext)
```

rkothiya commented:
/recheck smoke

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -6239,9 +6227,6 @@ pub_glfs_xreaddirplus_r(struct glfs_fd *glfd, uint32_t flags,
     GF_REF_GET(glfd);
 
-    GF_VALIDATE_OR_GOTO(this->name, xstat_p, out);
-    GF_VALIDATE_OR_GOTO(this->name, res, out);
```
This is a public function. We don't control the callers, so argument check is a must.

rkothiya replied:
done

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -5401,8 +5390,6 @@ pub_glfs_fd_set_lkowner(struct glfs_fd *glfd, void *data, int len)
         goto invalid_fs;
     }
 
-    GF_VALIDATE_OR_GOTO(this->name, data, out);
```
This is a public function. We don't control the callers, so argument check is a must.

rkothiya replied:
done

Commit:
Reducing the number of call to THIS

Changes done to incorporate review comments.
1) Retained argument check for public functions.

Updates: #1683

Change-Id: Ib4348b860b3aa942a9d0b0887d7c5be76c0eded4
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/line-coverage/cli-negative-case-and-function-coverage.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/basic/afr/split-brain-healing.t
./tests/line-coverage/cli-negative-case-and-function-coverage.t
https://build.gluster.org/job/gh_centos7-regression/1360/

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -5092,6 +5085,7 @@ glfs_realpath_common(struct glfs *fs, const char *path, char *resolved_path,
                      gf_boolean_t warn_deprecated)
 {
     int ret = -1;
+    xlator_t *this = THIS;
```
`THIS` will probably be incorrect before `__GLFS_ENTRY_VALIDATE_FD()`.

Please check other places where this also happens.

rkothiya replied:
done

Commit:
Reducing the number of call to THIS

Changes done to incorporate review comments.
1) Initialize "this" after functions/macros which change
   the value of "THIS".

Updates: #1683

Change-Id: I24b74eb4f7826316b10c92b539daacffd04e1019
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

xhernandez commented on api/src/glfs-fops.c:
```diff
@@ -6319,10 +6326,11 @@ GFAPI_SYMVER_PUBLIC_DEFAULT(glfs_xreaddirplus_get_stat, 3.11.0)
 struct stat *
 pub_glfs_xreaddirplus_get_stat(struct glfs_xreaddirp_stat *xstat)
 {
+    xlator_t *this = THIS;
```
Please, don't blindly apply the comments and wait until someone else checks if everything is right. It's the third time I raise this issue in this patch: **`THIS` cannot be used in public functions before calling `__GLFS_ENTRY_VALIDATE_FD()`**.

In this particular case the function signature doesn't include the required fields to correctly initialize `THIS`, so the only good solution here is to replace `this->name` in line 6333 by a string constant.

I haven't checked if there are more cases like this one.

mykaul replied:
Hi @xhernandez - just out of curiosity:
```
    DECLARE_OLD_THIS;
    __GLFS_ENTRY_VALIDATE_FD(glfd, invalid_fs);
```

This is seen tens of times. And DECLARE_OLD_THIS is actually calling 'THIS' and assigning it to old_THIS.
- Is that wrong?
- It's clearly redundant to the same assignment done @ __GLFS_ENTRY_VALIDATE_FD macro.

xhernandez replied:
> Hi @xhernandez - just out of curiosity:
> 
> ```
>     DECLARE_OLD_THIS;
>     __GLFS_ENTRY_VALIDATE_FD(glfd, invalid_fs);
> ```
> 
> This is seen tens of times. And DECLARE_OLD_THIS is actually calling 'THIS' and assigning it to old_THIS.
> 
> * Is that wrong?

It's not wrong if `THIS` contains a meaningful value or `old_THIS` is not used anywhere else beyond `__GLFS_EXIT_FS`. However in most (all ?) of the cases this is done in public API functions, which means that they can be called from any thread, and given that `THIS` accesses a thread-local variable, it will most probably be uninitialized, so at the very minimum, those calls are unnecessary IMO.

> * It's clearly redundant to the same assignment done @ __GLFS_ENTRY_VALIDATE_FD macro.

The main issue here is that `old_THIS` is declared by `DECLARE_OLD_THIS` macro, so it's needed to avoid compilation errors when `__GLFS_ENTRY_VALIDATE_FD` is used (it assigns to `old_THIS` but doesn't declare it). Probably `DECLARE_OLD_THIS` should have been limited to declare the variable, without initializing it, but most of the code already depends on this right now, so we can't change it easily.

In this particular case, `DECLARE_OLD_THIS` is saving a potential garbage value of `THIS`, but since it's not used, even if unnecessary, it's not a big problem. The big problem is to take a potentially uninitialized `THIS` an use it inside the function, as in this case. That's why I wrote the comment.

mykaul replied:
In the case of priv_glfs_process_upcall_event, it's a private function, but still looks different than the others, who, from what I can tell, do call __GLFS_ENTRY_VALIDATE_FS / __GLFS_ENTRY_VALIDATE_FD (which assign THIS):
```

    DECLARE_OLD_THIS;

    gf_msg_debug(THIS->name, 0, "Upcall gfapi callback is called");

    __GLFS_ENTRY_VALIDATE_FS(fs, err);

```

This is the only one I've seen that is different, and thus we can probably move DECLARE_OLD_THIS to actually just declare.

xhernandez replied:
I'm not sure if the `gf_msg_debug()` shouldn't be moved after `__GLFS_ENTRY_VALIDATE_FS()` in this case.

Since `DECLARE_OLD_THIS` is defined in glusterfs/globals.h, I assumed it's used globally in the code, however it seems its usage is limited to gfapi and one instance in rpc-clnt.c, so maybe it's easier than I thought to modify it.

mykaul replied:
The rpc-clnt one should go away via https://github.com/gluster/glusterfs/pull/3146/commits/6a4df415cafa8fa055cd5211d8d19452bc10fd80
